--- 
title: "Credibility of a repertoire report"
author: ""
date: "Updated: `r date()`"
knit: "bookdown::render_book"
site: bookdown::bookdown_site
documentclass: book
bibliography: "references.bib"
biblio-style: apalike
link-citations: yes
description: "Analysis notebook"
output:
    bookdown::gitbook:
      css: repcred.css 
      config:  
         fig_caption: yes
         toc:
            scroll_highlight: yes
            collapse: subsection
            depth: 4            
         download: yes
         sharing: no
         keep_md: true
params:
   date: !r date()
   echo:  FALSE
   cache: FALSE
   rep:  !r system.file("extdata", "ExampleDb.tsv", package = "repcred")
   outdir: !r tempdir()
   genome_file: !r NULL
   full_or_basic: !r FALSE
   downsample: !r TRUE
---

```{r global-options, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(fig.width=7, fig.asp=0.7, 
                      echo=params$echo,cache=params$cache,
                      warning=FALSE, message=FALSE,
                      out_dir=params$outdir,
                      eval.opts = c('eval', 'echo', 'fig.height', 'fig.width'))

# Save params to an .RData file. Useful for debugging.
save(params, file=file.path(params$outdir,"params.RData"))

# Load libraries
library(repcred)
library(airr)
library(sumrep)
library(plotly)
library(stringr)
library(ggplot2)
library(data.table)
library(seqinr)
library(kableExtra)

# Configuration and default values
gene_data = NA

#Below variables are set to give the orange , green and red colours. These colours were used as it makes the text easier to read and are more muted than the regular red , orange and green.
green = "LightGreen"
amber = "NavajoWhite"
red = "pink"
params
```

#  Input parameters 

```{r input-parameters}
#SECTION 1
if(is.null(params$genome_file)){
  gene_data = FALSE
}else{
 gene_data <-  seqinr::read.fasta(params$genome_file, as.string = TRUE,whole.header = TRUE)   
}

if(params$full_or_basic == "Full sumrep stats"){
   full_check=TRUE
}else{
   full_check = FALSE
}
params
```


```{r warning=FALSE}
repertoire <-fread(params$rep)
# Downsampling unless check box unticked 
num_keep <- 5000
if (params$downsample & nrow(repertoire) > num_keep) {
    repertoire <- repertoire[sample(nrow(repertoire), size=num_keep), ]
}


#Set up section colours
section_1=green
section_2=green
section_3=green
section_4=green
section_5=green
section_6=green
section_7=green
section_8=green
section_9=green
section_10=green
section_11=green

```
# Quality Control Stats

```{r warning=TRUE,results="asis"}
#SECTION 2
is_compliant <- validate_airr(repertoire)
missing_columns <- findMissingColumns(repertoire)
if (length(missing_columns) > 0) {
    writeLines("<h3> Columns with missing / No data </h3> ")
    kbl(data.table(column_name = missing_columns))
}
print(params$full_or_basic)
if (!is_compliant) {
 formattedErrorMeassage("The repertoire is not AIRR compliant. Exiting analysis.")
 knitr::knit_exit()   
}
```


# Non-nucleotides in sequence

```{r warning=FALSE}
#SECTION 3
check_nucleotides(repertoire)
```

# Statistics
Below provides useful statistics about your repertoire:

```{r warning=TRUE}
#SECTION 4
prod_data <-getFullRatioNonProductive(repertoire)
values = c(prod_data[[1]],prod_data[[2]],prod_data[[3]])
labels = c("Productive","Non-productive","NA/Not specified")
prod_info = data.table(a=values[[1]],b=values[[2]],c=values[[3]])
setnames(prod_info,labels)
barplot(values,names.arg = labels,col='light blue',ylab="Occurences" ,main="Number of productive vs non-productive sequences in repertoire file" )
kbl(prod_info)
```

```{r warning=TRUE}
if(getFullRatioNonProductive(repertoire)$non_prod != 0){
 ambig_non_prod_data <- fullCheckNonProductiveSeqs(repertoire)
print(ambig_non_prod_data)
kbl(ambig_non_prod_data)
 total_num=prod_data[[2]]
values = c(ambig_non_prod_data[[1]],ambig_non_prod_data[[2]],ambig_non_prod_data[[3]],total_num)

labels = c("Contain non-coding bases","Contain stop codons","Other non-normal sequences","Total number non-productive")
non_prod_table= data.table(a=values[[1]],b=values[[2]],c=values[[3]],d=total_num)
setnames(non_prod_table,labels)
par(mar=c(8, 13 ,5 ,3))
barplot(values,names.arg = labels,col="light blue",xlab="Occurences" ,main="Non-productive sequence issues overview",cex.names=1 ,horiz=TRUE,las=2)
knitr::kable(as.data.table(ambig_non_prod_data),col.names = labels)
# 
 kbl(non_prod_table) %>% kableExtra::column_spec(1,color="white",background=getPercColor(total_num,values[[1]])) %>% column_spec(2,color="white",background=getPercColor(total_num,values[[2]])) %>% column_spec(3,color="white",background=getNormalSeqNumColor(total_num,values[[3]]))
}else{
   print("There were no non-productive sequences present in the dataset")
}


```

```{r warning=FALSE,results="asis"}

sequence_length_data <- getSequenceLengths(repertoire)
hist(sequence_length_data,main="Histogram of sequence lengths" , xlab = "Sequence Length")
num_seqs_with_noncoding_vals <- getNumAmbiguousSeqs(repertoire)
kbl(getCoreStats(sequence_length_data))


#getSequenceCoverage(repertoire,gene_data)

#clusters=findAmplificationAreas(repertoire)
#knitr::kable(as.data.table(clusters),col.names = labels)
```
\n
Percentage of sequences where the V and J region are in-frame:
```{r , comment=NA,results="asis"}
if("vj_in_frame" %in% colnames(repertoire)&!any(is.na(repertoire$vj_in_frame))){
writeLines(paste(getInFramePercentage(repertoire),"%"))
}else{
   writeLines("Missing Column vj_in_frame")
}
```

# Gene Usage Statistics

```{r warning=TRUE,results="asis",fig.width=9}
#SECTION 5
genes_present <- getGeneData(repertoire)
#kbl(genes_present) 
```


```{r warning=TRUE,fig.height=20,results="asis"}
if(is.list(gene_data)){

fasta_genes <- readInGeneNamesIMGTFasta(gene_data)
gene_freq_table = geneCount(genes_present,fasta_genes)
freq_data_table=gene_freq_table
gene_freq_table=gene_freq_table[gene_freq_table$gene_count>0,]
gene_tables=getUniqueGenes(gene_freq_table)
genes_only_freq_table = gene_tables[[1]]

cat("\n")
print(kable(gene_freq_table))
cat("\n")

number_unqiue_alleles = length(gene_freq_table$gene_name)
number_unqiue_genes = length(genes_only_freq_table$gene_name)
labels=c("Number of unique alleles" , "Number of unique genes")
values=c(number_unqiue_alleles,number_unqiue_genes)

barplot(values,names.arg = labels,col='light blue',ylab="Counts" ,main="Comparison of unique genes to unique alleles" )
print(number_unqiue_genes)
print(ggplot(gene_freq_table,aes(x=gene_name,y=gene_count ,fill=gene_count))+geom_bar(stat="identity")+guides(fill = "none")+labs(x="Gene",y="Number of occurences")+coord_flip())


cat("\n")
print(kable(gene_tables[[1]]))
cat("\n")
allele_link=gene_tables[[2]]
gene_allele_freq_table=data.frame(table(allele_link$gene,allele_link$allele))
gene_allele_freq_table = gene_allele_freq_table[gene_allele_freq_table$Freq>0,]
gene_allele_freq_table$Var1 <- as.factor(gene_allele_freq_table$Var1)
print(ggplot(gene_allele_freq_table,aes(x=Var1,y=Freq,fill=Var2))+geom_bar(stat="identity")+guides(fill = "none", x=guide_axis(angle=90))+labs(x="Gene",y="Number of alleles present"))

cat("\n")
print(kable(allele_link))
cat("\n")

writeLines("<h3>Genes not present in the repertoire:</h3>")
print(kable(getAbsentGeneList(freq_data_table)))

}else{
   writeLines("No reference genome uploaded.")
}

```


# General Sumrep Statistics

```{r warning=TRUE,results="asis"}
#SECTION 6
plot(0:1,xaxt='n',yaxt='n',ann=FALSE)
legend("center",c("Mean +- standard dev","5% and 95% quantile points","Min and Max Value points"),fill=c("Dark blue","light blue","red"))

if("sequence_alignment" %in% colnames(repertoire)&!any(is.na(repertoire$sequence_alignment))){
hotspotCountDist(repertoire)
coldspotCountDist(repertoire)
gcContentDistribution(repertoire)
}else{
   section_6 = amber
    writeLines("Missing column : sequence_alignment \n Unable to run statistics : getHotspotCountDistribution , getColspotCountDistribution, getGCContentDistribution")
}


if("sequence_alignment" %in% colnames(repertoire) & "germline_alignment" %in% colnames(repertoire) &  !any(is.na(repertoire$sequence_alignment))& !any(is.na(repertoire$germline_alignment))){
 positionDistancesBetweenMutationDistribution(repertoire)
 distanceFromGermlineToSequenceDistribution(repertoire)  
}else{
   if(section_6 == amber){
      section_6 = red
   }else{
      section_6 = amber
   }
   writeLines("Missing columns : sequence_alignment , germline_alignment  \n Unable to run statistics : positionDistancesBetweenMutationDistribution , distanceFromGermlineToSequenceDistribution")
}

```


#  Pairwise Statistics

```{r warning=TRUE,results="asis"}
#SECTION 7
if(isTRUE(full_check)){
 plot(0:1,xaxt='n',yaxt='n',ann=FALSE)
legend("center",c("Mean +- standard dev","5% and 95% quantile points","Min and Max Value points"),fill=c("Dark blue","light blue","red"))

if("sequence_alignment" %in% colnames(repertoire)&!any(is.na(repertoire$sequence_alignment))){
CDR3pairwiseDistanceInfo(repertoire)
pairwiseDistDistribution(repertoire)
nearestNeighbourDistInfo(repertoire)
}else{
   section_7 = red
   writeLines("Missing column : Sequence_alignment \n Unable to run statistics")
}
}else{
   writeLines("Only the basic statistics were run.Please see other sections for the other statistics.")
}


```

#  Physiochemical Statistics

```{r}
if (all(is.na(repertoire$junction_aa))) {
    print("Skipping this section: `juncation_aa` is empty.")
}
```


```{r warning=TRUE,results="asis",eval=!all(is.na(repertoire$junction_aa))}
#SECTION 8
plot(0:1,xaxt='n',yaxt='n',ann=FALSE)
legend("center",c("Mean +- standard dev","5% and 95% quantile points","Min and Max Value points"),fill=c("Dark blue","light blue","red"))


#polarityDistribution(repertoire)
#aromaticityDistribution(repertoire)
#acidityDistribution(repertoire)
#basicityDistribution(repertoire)
# bulkinessDistribution(repertoire)
# chargeDistribution(repertoire)
aliphaticDistribution(repertoire)
print(mean(getAliphaticIndexDistribution(repertoire)),nm.rm=TRUE)
GRAVYDistribution(repertoire)

```


#  Insertion Length Distributions

```{r warning=TRUE,results="asis"}
#SECTION 9
plot(0:1,xaxt='n',yaxt='n',ann=FALSE)
legend("center",c("Mean +- standard dev","5% and 95% quantile points","Min and Max Value points"),fill=c("Dark blue","light blue","red"))
if("np1_length" %in% colnames(repertoire)&!any(is.na(repertoire$np1_length))){
 VJinsertionLengthDistribution(repertoire)
 VDinsertionLengthDistribution(repertoire)
}else{
 writeLines("Missing Column : np1_length")  
}
if("np2_length" %in% colnames(repertoire)&!any(is.na(repertoire$np2_length))){
DJinsertionLengthDistribution(repertoire)
}else{
   writeLines("Missing Column : np2_length")  
}
```

#  Prime deletion Distributions Statistics

```{r warning=TRUE,results="asis"}
#SECTION 10

plot(0:1,xaxt='n',yaxt='n',ann=FALSE)
legend("center",c("Mean +- standard dev","5% and 95% quantile points","Min and Max Value points"),fill=c("Dark blue","light blue","red"))

missing_col_count = 0
##
if("v_3p_del" %in% colnames(repertoire)&!any(is.na(repertoire$v_3p_del))){
 VGene3PrimeDeletionLengthDistribution(repertoire)  
missing_col_count = missing_col_count +1
}else{
   writeLines("\n Missing column : v_3p_del , Cannot run statistic \n")
}
if("v_5p_del" %in% colnames(repertoire)&!any(is.na(repertoire$v_5p_del))){
 VGene5PrimeDeletionLengthDistribution(repertoire) 
   missing_col_count = missing_col_count +1
}else{
   writeLines("Missing column : v_5p_del , Cannot run statistic \n")
}
##
if("d_3p_del" %in% colnames(repertoire)&!any(is.na(repertoire$d_3p_del))){
DGene3PrimeDeletionLengthDistribution(repertoire)
   missing_col_count = missing_col_count +1
}else{
   writeLines("Missing column : d_3p_del , Cannot run statistic \n")
}
if("d_5p_del" %in% colnames(repertoire)&!any(is.na(repertoire$d_5p_del))){
DGene5PrimeDeletionLengthDistribution(repertoire)
   missing_col_count = missing_col_count +1
}else{
   writeLines("Missing column : d_5p_del , Cannot run statistic \n")
}
##
if("j_3p_del" %in% colnames(repertoire)&!any(is.na(repertoire$j_3p_del))){
JGene3PrimeDeletionLengthDistribution(repertoire)
   missing_col_count = missing_col_count +1
}else{
   writeLines("Missing column : j_3p_del , Cannot run statistic \n")
}
if("j_5p_del" %in% colnames(repertoire)&!any(is.na(repertoire$j_5p_del))){
JGene5PrimeDeletionLengthDistribution(repertoire)
   missing_col_count = missing_col_count +1
}else{
   writeLines("Missing column : j_5p_del , Cannot run statistic \n")
}

if(missing_col_count > 3){
   section_10=red
}
if(missing_col_count > 0){
   section_10=amber
}
```



#  Possible Chimerisms
Below table shows the comparison between the total number of sequences compares to the total number of unique CDR3 sequences and then compared to the number of CDR3 sequences that have multiple different v-call genes associated with them.
```{r CDR3_Chimera_Check, warning = TRUE ,fig.width=9}
#SECTION 11
if(full_check == TRUE){
repertoire = repertoire[!is.na(repertoire$sequence),]
cdr3_seq_info = checkCDR3(repertoire)

#table(cdr3_seq_info$cdr3_seq)
cdr3_vcalls = getVCalls(cdr3_seq_info,repertoire,FALSE)
total_num_unique_seq = length(unique(repertoire$sequence))
##
num_uniq_cdr3_seq = length(unique(cdr3_seq_info$cdr3_seqs))

num_occur_multiple_gene_call <- length(which(data.frame(table(cdr3_seq_info$cdr3_seqs))$Freq > 1))
#kbl(cdr3_vcalls)

####
# 
 freq_table= data.frame(table(as.character(cdr3_vcalls$seq),as.character(cdr3_vcalls$v_call_genes)))
 freq_table= freq_table[freq_table$Freq >1,]
multiple_vgene_freq = data.frame(table(as.character(freq_table$Var1)))
multiple_vgene_freq = multiple_vgene_freq[multiple_vgene_freq$Freq>1,]
multiple_gene_calls = length(table(freq_table$Var1))
#print(freq_table)
seqs_stats_vals=c(total_num_unique_seq,num_uniq_cdr3_seq,multiple_gene_calls)
labels = c("Unique complete seqs" , "Unique CDR3 seqs" , "Unique CDR3 seqs with multiple single v-calls" )

##
par(mar=c(8, 13 ,5 ,3))
barplot(seqs_stats_vals,names.arg = labels,col='dark blue',xlab="Occurences" ,main="CDR3 Sequence comparisons", las=2, horiz=TRUE ,cex.names=0.8)


freq_table$Var2 <- as.factor(freq_table$Var2)
ggplot(freq_table,aes(x=Var2,y=Freq,fill=Var1))+geom_bar(stat="identity")+guides(fill = "none", x=guide_axis(angle=90))+labs(x="Gene",y="Number of sequences present")
}else{
   writeLines("Only Basic stats were run , if you would like to get stats of possible chimeric sequences please run full stats.")
   section_11 = "DarkGrey"
}
```


Below contain CDR3 sequences with more than one v_call:

```{r warning = TRUE, results="asis" } 
options(width = 2000)
num_to_display=0
if(length(repertoire$sequence)>100000){
   num_to_display = 3
}else{
   num_to_display = 6
}

# if(seq_type == "")
# display_type
if (full_check == TRUE){
plotVgeneDist(cdr3_vcalls,num_to_display, "aa",multiple_vgene_freq)}
#kbl(most_chimeric)
```

```{r warning = TRUE, results="asis"}
#kbl(as.data.table(findAmplificationAreas(repertoire)))

addTrafficLighting(c(section_1,section_2,section_3,section_4,section_5,section_6,section_7,section_8,section_9,section_10,section_11))
```


